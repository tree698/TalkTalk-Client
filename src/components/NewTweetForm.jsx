import React, { useEffect, useState } from 'react';
import { FaPaperPlane } from 'react-icons/fa';
import { useLocation } from 'react-router-dom';
import { useApiContext } from '../context/ApiContext';
import toast from 'react-hot-toast';

export default function NewTweetForm() {
  const [tweet, setTweet] = useState('');
  const [error, setError] = useState('');

  const { tweetService } = useApiContext();
  const {
    state: { id },
  } = useLocation();

  useEffect(() => {
    error && toast.error(error);
  }, [error]);

  const handleSubmit = async (event) => {
    event.preventDefault();
    tweetService
      .createTweet(tweet, id)
      .then((created) => {
        setTweet('');
      })
      .catch((error) => {
        setError((prev) => error.toString());
        setTimeout(() => {
          setError('');
        }, 10000);
      });
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex justify-between items-center mb-2 text-xl"
    >
      <input
        type="text"
        placeholder="Type a talk"
        value={tweet}
        required
        autoFocus
        onChange={(e) => setTweet((prev) => e.target.value)}
        className="w-full px-3 my-3 py-3 border-2 border-lightGray outline-none rounded-md placeholder:text-lightGray placeholder:italic shadow-lg"
      />
      <button
        type="submit"
        className="w-12 h-12 flex justify-center items-center rounded-full ml-3 text-2xl text-white bg-lightGray hover:scale-110 hover:brightness-110 transition-all delay-150 duration-300 ease-in-out"
      >
        <FaPaperPlane className="" />
      </button>
    </form>
  );
}
