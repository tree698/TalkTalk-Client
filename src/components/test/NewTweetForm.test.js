import NewTweetForm from '../NewTweetForm';
import renderer from 'react-test-renderer';
import { withApiContext, withRouter } from '../../test/utils';
import { Route } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('NewTweetForm', () => {
  let fakeContext;
  beforeEach(
    () =>
      (fakeContext = {
        tweetService: jest.fn(),
      })
  );

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<NewTweetForm />} />, {
          state: { id: 'fakeId' },
        }),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('type tweet correctly', () => {
    render(
      withApiContext(
        withRouter(<Route path="/" element={<NewTweetForm />} />, {
          state: { id: 'fakeId' },
        }),
        fakeContext
      )
    );

    userEvent.type(screen.getByRole('textbox'), 'fakedTweet');

    const inputElement = screen.getByRole('textbox');
    const inputValue = inputElement.value;
    expect(inputValue).toBe('fakedTweet');
  });
});
