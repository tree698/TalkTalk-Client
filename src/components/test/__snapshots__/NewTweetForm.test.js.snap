// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTweetForm renders correctly 1`] = `
<form
  className="flex justify-between mt-3 md:mt-4 lg:mt-5 border border-superLightGray rounded-md overflow-hidden shadow-inner"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="w-full px-2 py-1 md:py-2 outline-none placeholder:text-lightGray placeholder:italic placeholder:text-xs md:placeholder:text-sm lg:placeholder:text-base"
    onChange={[Function]}
    placeholder="Type a talk"
    required={true}
    type="text"
    value=""
  />
  <button
    className="text-xl md:text-2xl lg:text-3xl px-1 md:px-2 lg:px-3"
    type="submit"
  >
    <svg
      className="text-lightGray hover:text-slate500 hover:scale-105"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 4H6c-1.103 0-2 .897-2 2v5h2V8l6.4 4.8a1.001 1.001 0 0 0 1.2 0L20 8v9h-8v2h8c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2zm-7 6.75L6.666 6h12.668L13 10.75z"
      />
      <path
        d="M2 12h7v2H2zm2 3h6v2H4zm3 3h4v2H7z"
      />
    </svg>
  </button>
</form>
`;
